Windows: set main thread name without re-encoding

ChrisDenton opened the pull request:
As a minor optimization, we can skip the runtime UTF-8 to UTF-16 conversion.

ChrisDenton committed:
Windows: set main thread name without reencoding

workingjubilee commented on library/std/src/sys/pal/windows/mod.rs:
```diff
@@ -60,7 +60,7 @@ pub unsafe fn init(_argc: isize, _argv: *const *const u8, _sigpipe: u8) {
     // Normally, `thread::spawn` will call `Thread::set_name` but since this thread already
     // exists, we have to call it ourselves.
-    thread::Thread::set_name(&c"main");
+    thread::Thread::set_name_wide(&['m' as u16, 'a' as u16, 'i' as u16, 'n' as u16, 0]);
```
It bugs me that `b"main\0".map(|b| b as u16)` works but `c"main".map(|b| b as u16)` doesn't.

ChrisDenton replied:
Hm, I should probably make a macro to do const encoding of `&str` to null terminated `&[u16]`.

ChrisDenton replied:
I went ahead and made a little macro for const utf16 strings. I just did a pretty direct translation of the wikipedia articles.

ChrisDenton committed:
Add const UTF-8 to UTF-16 conversion macros

`wide_str!` creates a null terminated UTF-16 string whereas `utf16!` just creates a UTF-16 string without adding a null.

jieyouxu commented on library/std/src/sys/pal/windows/api.rs:
```diff
@@ -34,6 +34,97 @@ use core::ptr::addr_of;
+/// Const convert UTF-8 to UTF-16, for use in the wide_str macro.
+///
+/// Note that this is designed for use in const contexts so is not optimized.
+pub const fn to_utf16<const UTF16_LEN: usize>(s: &str) -> [u16; UTF16_LEN] {
```
Do you think if this needs a test? This feels a bit non-trivial.

ChrisDenton replied:
Yeah, would be best. Also I should probably add a null check somewhere.

ChrisDenton replied:
I've added a few test cases that should exercise the different code paths. Though in practice I suspect it won't get much use beyond ASCII. Well, possibly a bit of latin-1.

workingjubilee commented on library/std/src/sys/pal/windows/api.rs:
```diff
@@ -34,6 +34,105 @@ use core::ptr::addr_of;
+                output[pos] = s[i] as u16;
+                i += 1;
+                pos += 1;
+            }
+            2 => {
+                // Bits: 110xxxxx 10xxxxxx
+                output[pos] = ((s[i] as u16 & 0b11111) << 6) | (s[i + 1] as u16 & 0b111111);
+                i += 2;
+                pos += 1;
+            }
+            3 => {
+                // Bits: 1110xxxx 10xxxxxx 10xxxxxx
+                output[pos] = ((s[i] as u16 & 0b1111) << 12)
+                    | ((s[i + 1] as u16 & 0b111111) << 6)
+                    | (s[i + 2] as u16 & 0b111111);
+                i += 3;
+                pos += 1;
+            }
+            4 => {
+                // Bits: 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+                let mut c = ((s[i] as u32 & 0b111) << 18)
+                    | ((s[i + 1] as u32 & 0b111111) << 12)
+                    | ((s[i + 2] as u32 & 0b111111) << 6)
+                    | (s[i + 3] as u32 & 0b111111);
+                // re-encode as UTF-16 (see https://en.wikipedia.org/wiki/UTF-16)
+                // - Subtract 0x10000 from the code point
+                // - For the high surrogate, shift right by 10 then add 0xD800
+                // - For the low surrogate, take the low 10 bits then add 0xDC00
+                c -= 0x10000;
+                output[pos] = ((c >> 10) + 0xD800) as u16;
+                output[pos + 1] = ((c & 0b1111111111) + 0xDC00) as u16;
+                i += 4;
+                pos += 2;
+            }
+            // valid UTF-8 cannot have any other values
+            _ => unreachable!(),
+        }
+    }
+    output
+}
```
nice work. I feel like at least some of this should be using more public std API instead of a bunch of sorcerous isopsephia, but I looked for equivalents and couldn't find any in the stdlib, so this will do for now.

workingjubilee commented on library/std/src/sys/pal/windows/api.rs:
```diff
@@ -34,6 +34,105 @@ use core::ptr::addr_of;
+            n => n as usize,
+        };
+        i += utf8_len;
+        len += if utf8_len < 4 { 1 } else { 2 };
```
hmm. this feels slightly wrong. the 3-byte case...
1110xxxx 10xxxxxx 10xxxxxx?

that's 16 bits of raw entropy, but if memory serves, not all u16s are individually valid in UTF-16? because of the surrogate reservation? so there are some 3-byte UTF-8 sequences that are two u16s, aren't there? there's some more nuances of the encoding that I'm not quite remembering, and I know not all 3-byte sequences are valid either, but...

if I'm wrong, this needs a comment explaining why I'm wrong, since it seems like a "naive" bit of code, and a sophomoric programmer might try to "correct" it otherwise.

ChrisDenton replied:
Those are also not valid UTF-8. I'll add a comment. See https://en.wikipedia.org/wiki/UTF-8#Invalid_sequences_and_error_handling

> Since RFC 3629 (November 2003), the high and low surrogate halves used by [UTF-16](https://en.wikipedia.org/wiki/UTF-16) (U+D800 through U+DFFF) and code points not encodable by UTF-16 (those after U+10FFFF) are not legal Unicode values, and their UTF-8 encoding must be treated as an invalid byte sequence. Not decoding unpaired surrogate halves makes it impossible to store invalid UTF-16 (such as Windows filenames or UTF-16 that has been split between the surrogates) as UTF-8,[[e]](https://en.wikipedia.org/wiki/UTF-8#cite_note-PEP_529_quote-33) while it is possible with [WTF-8](https://en.wikipedia.org/wiki/UTF-8#WTF-8).

ChrisDenton replied:
I'll add a comment. UTF-16 can encode values U+0000 to U+D7FF and U+E000 to U+FFFF in [a single u16](https://en.wikipedia.org/wiki/UTF-16#U+0000_to_U+D7FF_and_U+E000_to_U+FFFF).

The hole in the middle are for the surrogate pairs but these [aren't valid UTF-8](https://en.wikipedia.org/wiki/UTF-8#Invalid_sequences_and_error_handling) either:

> Since RFC 3629 (November 2003), the high and low surrogate halves used by [UTF-16](https://en.wikipedia.org/wiki/UTF-16) (U+D800 through U+DFFF) and code points not encodable by UTF-16 (those after U+10FFFF) are not legal Unicode values, and their UTF-8 encoding must be treated as an invalid byte sequence. Not decoding unpaired surrogate halves makes it impossible to store invalid UTF-16 (such as Windows filenames or UTF-16 that has been split between the surrogates) as UTF-8,[[e]](https://en.wikipedia.org/wiki/UTF-8#cite_note-PEP_529_quote-33) while it is possible with [WTF-8](https://en.wikipedia.org/wiki/UTF-8#WTF-8).

workingjubilee replied:
...The answer to this bugs me on some level because the encoding could be more compact, I think, even in the presence of the surrogate pair codepoints, by offsetting the encoding ranges by the fixed number of the last range, even without any further cleverness.

ChrisDenton replied:
Time to propose a new Unicode encoding?

workingjubilee replied:
Ah, so now I understand. This problem is the problem of overlong encodings. They were once accepted, so we have a large number of unused bits.

workingjubilee commented on library/std/src/sys/pal/windows/api.rs:
```diff
@@ -34,6 +34,105 @@ use core::ptr::addr_of;
+
+/// Creates a UTF-16 string from a str without null termination.
+macro_rules! utf16 {
+    // Note: this macro uses triple underscores to avoid const cycles
+    ($str:expr) => {{
+        const ___UTF8: &str = $str;
+        const ___UTF16_LEN: usize = crate::sys::pal::windows::api::utf16_len(___UTF8);
+        const ___UTF16: [u16; ___UTF16_LEN] = crate::sys::pal::windows::api::to_utf16(___UTF8);
+        &___UTF16
+    }};
```
huh. const cycles...?

ChrisDenton replied:
I mean like:

```rust
const UTF8: &str = "hello";
{
    const UTF8: &str = UTF8;
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=66187861d2fa095ad67c03359b6d097b)

ChrisDenton replied:
That would be sensible but... https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85373b84b46ca4d49dd9544030fdc081

workingjubilee replied:
Oh, so it's only a problem if we add it as an argument.

ChrisDenton replied:
Yeah. I guess there's the option of saying "don't do that" or requiring it to be a literal. But the workaround is relatively painless as macros go.

workingjubilee commented on library/std/src/sys/pal/windows/api/tests.rs:
```diff
@@ -0,0 +1,14 @@
+fn test_utf16_macros() {
+    check_utf16!("hello world");
+    check_utf16!("€4.50");
+    check_utf16!("𨉟呐㗂越");
```
chào

ChrisDenton committed:
Add comment on UTF-16 surrogates

workingjubilee reviewed (approved):
r=me with comment

ChrisDenton commented:
@bors r=workingjubilee

bors commented:
:pushpin: Commit 614e793013c75163bf7497ba2eb9f66677cb00e6 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 614e793013c75163bf7497ba2eb9f66677cb00e6 -->
<!-- homu: {"type":"Approved","sha":"614e793013c75163bf7497ba2eb9f66677cb00e6","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
Okay, following fmease's explanation I think using a decl macro would be fine since we're std and get to use nightly features when we want:
@bors r-

workingjubilee commented:
Yay! ( I don't mean to be annoying, I just don't think we should embrace a fragile proliferation of underscores if we don't have to. )

@bors r+

bors commented:
:pushpin: Commit 19f04a7d6878fc4c258ba3d4374e81c8bbeca2e0 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 19f04a7d6878fc4c258ba3d4374e81c8bbeca2e0 -->
<!-- homu: {"type":"Approved","sha":"19f04a7d6878fc4c258ba3d4374e81c8bbeca2e0","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
